File: 141.cpp
/*
very classical problem . 
Very high chance that in interview you will face this question .
So try to see this algorithm from every angle like .... 
time complexity 
space complexity
edge cases
on which node we will have the cycle start if we start traversing the list from the given head .
*/
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if(head == NULL){
            return false ; 
        }
        ListNode * shortStep = head , *longStep = head->next ;
        while(shortStep != longStep){
            if(longStep){
                longStep = longStep->next  ;
                if(longStep){
                    longStep = longStep->next ;
                    shortStep = shortStep->next ; 
                }
                else{
                    return false ;
                }
            }
            else{
                return false ;
            }
        }
        return true ; 
    }
};

File: python_solutions/141.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        # Solution overview
        # we can use Floyd's cycle detection algorithm

        slow_head = head
        fast_head = head
        while fast_head is not None and fast_head.next is not None and fast_head.next.next is not None:
            slow_head = slow_head.next
            fast_head = fast_head.next.next
            if slow_head == fast_head:
                return True
        return False
        
